broadinstitute/single_cell_portal_core
# Creating anonymous publication reviewer access system (SCP-3574)

Currently, granting access to reviewers from scientific journals to private SCP studies is mostly broken.  It relies on creating shared Google logins that multiple reviewers attempt to authenticate into SCP with.  This fails to work most of the time as Google's security rightly flags the multiple logins as suspicious, and employs 2-factor authentication to validate that the reviewer does in fact have access.  Reviewers are not allowed to have direct contact with corresponding authors, therefore they will not have access to any of the second factors.  This means login is blocked, and review is impossible.  

This update adds the `ReviewerAccess` and `ReviewerAccessSession` models to facilitate anonymous reviewer access to private studies.  The flow works as follows: a study owner visits the "Study Settings" tab for their private study, and enables reviewer access.  Once enabled, the study owner is provided with an access URL and a PIN code, which they give to their corresponding journal editor.  When reviewers are assigned, the editor provides them with the credentials, and they navigate to the access URL and enter the PIN code.  Successful authentication via the PIN creates a signed session cookie which allows limited access to the private study (reviewers can interact with visualizations, but not download data or visit the Terra workspace).  These sessions last for 4 hours, at which point they expire and cannot be used again.  The session cookies themselves only apply to a single study, meaning a reviewer can have access to multiple private studies (they are name-spaced with the study accession).  Furthermore, these are "signed" cookies, meaning that they are signed by the application and cannot be altered in any way.  Attempts to do so will cause the cookie to delete itself.

A reviewer can request a new session by re-entering the PIN at the access URL.  Study owners can also reset credentials to generate a new access URL & PIN, which will also invalidate all active reviewer sessions.  Furthermore, the reviewer access functionality itself has an expiration date, at which point all reviewer access is blocked.  Study owners can also manually turn off access at any time.

MANUAL TESTING
1. On a private study with visualizations enabled, navigate to the "Study Settings" tab, select "Yes" for "Access Enabled?", then click "Update Study Settings"
![Screen Shot 2021-08-18 at 1 16 30 PM](https://user-images.githubusercontent.com/729968/129942933-d1b77e0c-2493-48d7-8fc0-327260ab9e21.png)
2. Once saved, click the "Copy to clipboard" button under "Current Access Credentials" and paste into a text document to see the credentials.
![Screen Shot 2021-08-18 at 1 12 14 PM](https://user-images.githubusercontent.com/729968/129942395-1d8a14b1-0c14-436a-be66-3b35ae482299.png)
3. Paste the access URL into a **new** browser window (preferably incognito), then on the reviewer access page, enter the 10 digit PIN code and click "Submit"
![Screen Shot 2021-08-18 at 1 17 46 PM](https://user-images.githubusercontent.com/729968/129943131-435e8bf0-69f1-446f-87e6-eda5f14b0234.png)
4. After submission, you should be redirected to the study overview page.  Click around the study, trying different cluster/annotation and gene search combinations
5. Open the Chrome Dev Tools window, and click into the "Application" tab, then select "Cookies" from the left.  You should see an entry for `reviewer_session_[:accession]` in the list, like so:
![Screen Shot 2021-08-18 at 1 06 14 PM](https://user-images.githubusercontent.com/729968/129941588-da6b8fc5-9024-4265-bb57-ef136d26178c.png)
6. (Optional) In the original window, select either "Yes" for "Reset Access Credentials" or "No" for "Access Enabled" and save (both will invalidate the current session)
7. (Optional) Back in the incognito window, refresh the page and note that you are redirected to the homepage with a notice that either the session has expired (if you rotated the credentials) or that you do not have permission to perform the action (if you revoked access altogether)

This PR satisfies SCP-3574.
